Shader Blueprint class for a unified shader language

marauder2k7 opened the pull request:
- ShaderBlueprint class
- Basic unified shader language parser that exports both hlsl and glsl files 
- Simple example shaderblueprint included.

marauder2k7 committed:
Initial commit

-ShaderBlueprint class
-Basic unified shader language parser that exports both hlsl and glsl files
-Simple example shader included.

marauder2k7 committed:
Hook up function

-Now functions should export
-SamplerTypeToStringGLSL now more universal for use with args and uniforms.
-Add copyright notice

marauder2k7 committed:
Update shaderBlueprint.cpp

forgot spaces after in / out data type

lukaspj commented:
At a glance, two points:

1. A documentation of the spec for these blueprint files would be nice. It can be added as a Doxygen comment like [this](https://github.com/TorqueGameEngines/Torque3D/blob/development/Engine/source/console/consoleObject.h#L114-L176)
2. I don't think ".blueprint" is a good extension, it's too generic. I'd suggest either ".tsl" (Torque shader language), ".usl" (unified shader language), ".shader" or something similar.

marauder2k7 committed:
Change file type

-Change file type to .tsl

marauder2k7 committed:
change type to tlsl

-TLSL FTW!

marauder2k7 committed:
basicClouds working

-Working example of basicClouds using shaderBlueprint instead of shaderData

-BUG: gl side does not render correctly.

marauder2k7 committed:
correctSSP functionality

-add #pragma SCREENSPACEPOS to fix screenspace position in gl

marauder2k7 committed:
More functionality

-Added HDRENCODE pragma to use hdrEncode functions
-Added macros can be defined with #define in each shader section.
-GL side now adds global macros at compile time.
-Shader functions were having issues with first argument.
-Added TANGENT shader semantic

marauder2k7 committed:
CloudLayer uses ShaderBlueprint

-Added layout binding for glsl output (Thanks Jeff 😍)
-CloudLayer now uses a shader blueprint instead of shaderData

marauder2k7 committed:
Global var tracking

-Added ability to define global vars inside shader types
-Moved the function reader off to its own function for reuse when we handle includes.

marauder2k7 committed:
Update shaderBlueprint.cpp

-Added new function for processing in function lines, if for while do else endif all checked for
-Added check to make sure fileExtension is tlsl

marauder2k7 committed:
Update shaderBlueprint.cpp

-Added more types to the tables.
-Handle unroll  (needs expanded)
-Format single line if statements better (output needs to be readable for debugging)

marauder2k7 committed:
Implement shadercache

-Added shaderCache functionality from shaderData
-functionLine call needs to be after the conversion.

marauder2k7 committed:
Include file parsing

-Added ability to parse include files Note: these too need to be tlsl though they do not need any specifiers, could be just filled with functions and defines

-Added safeties around hlsl built in functions being translated to glsl

-Added clip function parsing (think clip will only be used in pixel shader?)

marauder2k7 committed:
Merge remote-tracking branch 'upstream/development' into CrossShaderCompilation

marauder2k7 committed:
Structs

-Added struct parsing for shaders (internal structs)
- Fixed arrays

marauder2k7 committed:
Shader struct function printing

-Shader struct functions were not being printed out.
-Shader function printing now comes with a start depth so we can format better
-Added GFXShaderConstType void for functions

marauder2k7 committed:
well... this was a lot simpler than i previously thought

-Removed a lot of unnecessary complexity.

marauder2k7 committed:
reimplement function

-ReAdded ability to parse functions

HLSL side needs to be able to change the function arguments to include texture and samplerstate, also anywhere the sampler argument is used will need to pass texture and sampler also.

marauder2k7 committed:
Update shaderBlueprint.cpp

marauder2k7 committed:
Added multi stage

-Added multi stage functionality:

Stages act kind of like passes for shaders, you can define multiple properties inside stage blocks that will be applied to the shader files when that stage is called for. for example in gaussian blur there is a define for blur_dir, before this would require multiple shaderData types to account for the change in blur direction between postfx stages. Now you define those stages inside a blueprint file and just select each one as it is needed.

marauder2k7 committed:
Update shaderBlueprint.cpp

marauder2k7 committed:
Update shaderBlueprint.cpp

-reading included files was trapped in an infinite loop
-arrays in glsl are initialized differently (may need to split this one out)
-due to certain initializations the main readshaderdata would be trapped in an infinite loop as well
